osquery/events/darwin/openbsm.cpp

```diff
@@ -41,54 +42,83 @@ void OpenBSMEventPublisher::tearDown() {
   }
 }
 
+void OpenBSMEventPublisher::acquireMessages() {
+  auto buffer = static_cast<unsigned char*>(nullptr);
+  auto reclen = au_read_rec(audit_pipe_, &buffer);
+  if (reclen <= 0) {
+    return;
+  }
+
+  // We'll use these to dequeue below.
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+
+  auto event_id = 0;
+  auto bytesread = 0;
+  while (bytesread < reclen) {
+    if (au_fetch_tok(&tok, buffer + bytesread, reclen - bytesread) == -1) {
+      break;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER32_EX:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER64:
+      event_id = tok.tt.hdr64.e_type;
+      break;
+    case AUT_HEADER64_EX:
+      event_id = tok.tt.hdr64_ex.e_type;
+      break;
+    }
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+
+  // We probably don't need a lambda here but it's useful to put debug
+  // lines in to validate destruction.
+  std::shared_ptr<unsigned char> sp_buffer(buffer,
+                                           [](unsigned char* p) { delete p; });
+
+  auto ec = createEventContext();
+  ec->event_id = event_id;
+  ec->tokens = tokens;
+  ec->buffer = sp_buffer;
+  fire(ec);
+}
+
 Status OpenBSMEventPublisher::run() {
   if (audit_pipe_ == nullptr) {
     return Status(1, "No open audit_pipe");
   }
-  tokenstr_t tok;
-  auto reclen = 0;
-  auto bytesread = 0;
-  auto event_id = 0;
-  auto buffer = static_cast<unsigned char*>(nullptr);
-  std::vector<tokenstr_t> tokens{};
 
-  while (!isEnding() && (reclen = au_read_rec(audit_pipe_, &buffer)) != -1) {
-    bytesread = 0;
+  fd_set fdset;
+  FD_ZERO(&fdset);
+  FD_SET(fileno(audit_pipe_), &fdset);
+  struct timeval timeout;
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 0;
```

Comment:
@uptycs-nishant, I changed this to be 0 so it is more obvious and explicit that we only want `::select` to check if there is data to be read.

Reply:
I problem still remains the same(i.e we might lose events).
scenario -
1. select returns, reporting no events.
2. breakout from the loop
3. go to sleep for 200ms.
4. Burst of events happen, internal queue got full, while we are sleeping, might result into event loss.

Reply:
So, I believe we should sleep on select and once the events arrive we should wake up and read the events without delaying (dont let the internal queue get full).

Reply:
I agree with you about the potential to overflow the queue. My ask is that we address this in a follow up PR since it is not straightforward to skip the 200ms imposed by the eventing run loop.

Step 1, add polling.
Step 2, protect against overflowing.